cluster/cluster_interface.ml

```diff
@@ -14,12 +14,26 @@ type debug_info = string [@@deriving rpcty]
 (** Name of the cluster *)
 type cluster_name = string [@@deriving rpcty]
 
-(** An IPv4 address (a.b.c.d) *)
-type address = IPv4 of string [@@deriving rpcty]
+type ip_with_hostname = {ip: string; hostname: string} [@@deriving rpcty]
 
-let printaddr () = function IPv4 s -> Printf.sprintf "IPv4(%s)" s
+(** An cluster member's address
+  * IPv4 should be avoided if possible, but is kept for backwards compatibility *)
+type address = IPv4 of string | IP_with_hostname of ip_with_hostname
```

Comment:
As I said in the xen-api PR, I think that using the name hostname here is leaking implementation details, I'd rather tell users of this interface that the "hostname" is a host uuid. How it's used in the cluster server is a different matter.

I was thinking along the lines of
```ocaml
type ip_and_uuid = {ip: Ipaddr.t; uuid: string} [@@deriving rpcty]

type address = IPv4 of string | Xapi_host of ip_and_uuid
```